native-image-tests/src/main/kotlin/okhttp3/RunTests.kt

```diff
@@ -17,60 +17,87 @@ package okhttp3
+import org.junit.platform.launcher.core.EngineDiscoveryOrchestrator
 import org.junit.platform.launcher.core.LauncherConfig
 import org.junit.platform.launcher.core.LauncherDiscoveryRequestBuilder
 import org.junit.platform.launcher.core.LauncherFactory
 import org.junit.platform.launcher.listeners.SummaryGeneratingListener
 import java.io.PrintWriter
 import kotlin.system.exitProcess
 
-val knownTests = listOf(
-  okhttp3.logging.IsProbablyUtf8Test::class.java,
-  okhttp3.logging.LoggingEventListenerTest::class.java,
-  okhttp3.logging.HttpLoggingInterceptorTest::class.java,
-  okhttp3.sse.internal.EventSourceHttpTest::class.java,
-  okhttp3.sse.internal.ServerSentEventIteratorTest::class.java,
-//  okhttp3.CallTest::class.java,
-//  okhttp3.internal.http.CancelTest::class.java,
-  SampleTest::class.java)
-
+/**
+ * Graal main method to run tests with minimal reflection and automatic settings.
+ * Uses the test list in native-image-tests/src/main/resources/testlist.txt.
+ */
 fun main() {
   System.setProperty("junit.jupiter.extensions.autodetection.enabled", "true")
 
+  val selectors = testSelectors()
+
   val summaryListener = SummaryGeneratingListener()
   val treeListener = treeListener()
 
-  val jupiterTestEngine = JupiterTestEngine()
+  val jupiterTestEngine = buildTestEngine()
 
   val config = LauncherConfig.builder()
     .enableTestExecutionListenerAutoRegistration(false)
     .enableTestEngineAutoRegistration(false)
     .enablePostDiscoveryFilterAutoRegistration(false)
     .addTestEngines(jupiterTestEngine)
-    .addTestExecutionListeners(summaryListener, treeListener, DotListener)
+    .addTestExecutionListeners(DotListener, summaryListener, treeListener)
     .build()
   val launcher: Launcher = LauncherFactory.create(config)
 
-  val request: LauncherDiscoveryRequest = LauncherDiscoveryRequestBuilder.request()
-    // TODO replace junit.jupiter.extensions.autodetection.enabled with API approach.
-//    .enableImplicitConfigurationParameters(false)
-    .selectors(knownTests.map { selectClass(it) })
-    .build()
+  val request: LauncherDiscoveryRequest = buildRequest(selectors)
 
-  val result = launcher.execute(request)
+  launcher.execute(request)
 
   val summary = summaryListener.summary
   summary.printTo(PrintWriter(System.out))
 
   exitProcess(if (summary.testsFailedCount != 0L) -1 else 0)
 }
 
+fun buildTestEngine(): TestEngine = JupiterTestEngine()
+
+fun testSelectors(): List<DiscoverySelector> {
+  val sampleTestClass = SampleTest::class.java
+  return sampleTestClass.getResource("/testlist.txt")
+    .readText()
+    .lines()
+    .filter { it.isNotBlank() }
+    .map {
+      selectClass(Class.forName(it, false, sampleTestClass.classLoader))
+    }
+}
+
+fun buildRequest(selectors: List<DiscoverySelector>): LauncherDiscoveryRequest {
+  val request: LauncherDiscoveryRequest = LauncherDiscoveryRequestBuilder.request()
+    // TODO replace junit.jupiter.extensions.autodetection.enabled with API approach.
```

Comment:
what’s this mean? Changing how we configure JUnit here? Or how we did it in our tests?

Reply:
https://github.com/junit-team/junit5/issues/2470

I'm this close to a config driven completely without any magic.  But then I nee to set a system property to make it use JAR services mechanism.  I'd prefer someway to do that in code.